### PR TITLE
image upload

### PR DESCRIPTION
**Pull request by randombenj**
Allow an image to be uploaded.
For this the image upload has to be enabled with `--upload true`

You can then push an image with curl:

```sh
curl -X POST \
  -F "image=@imate.tar.xz" \
  -F "aliases=rlx/latest" \
  -F "remote=https://localhost:8443" localhost:3000/api/v1/image
```

### CODE REVIEW

File: packages/api/src/lxd/lxd.service.ts [hunk 1]
```diff
@@ -30,12 +31,13 @@ export class LXDService {
       rejectUnauthorized: false
     };
 
-    if (this.settings.lxd && this.settings.lxd.key) {
-      options.key = this.settings.lxd.key;
-    }
-
-    if (this.settings.lxd && this.settings.lxd.cert) {
-      options.cert = this.settings.lxd.cert;
+    if (this.settings.lxd) {
```
Comment (BrunnerLivio) [inline packages/api/src/lxd/lxd.service.ts]: If you upgrade to TypeScript 3.7 you can also use this
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

File: packages/api/src/lxd/lxd.service.ts [hunk 2]
```diff
@@ -106,11 +108,99 @@ export class LXDService {
...
+    const operation = (await axios.post(`${remote}/1.0/images`, image.buffer, {
+      headers: {
+        'X-LXD-Public': '1',
+      },
+      maxContentLength: Infinity
+    })).data.operation;
+
+    const result = (await axios.get(`${remote}/${operation}/wait`)).data;
+    if (
+      result &&
```
Comment (BrunnerLivio) [inline packages/api/src/lxd/lxd.service.ts]: Also use optional chaining here

File: packages/api/src/lxd/lxd.service.ts [hunk 3]
```diff
@@ -106,11 +108,99 @@ export class LXDService {
...
+   * @param remote The remote to add the image to (format: https://localhost:8334)
+   */
+  async importImage(remote: string, image: any) {
+    const axios = this.getAxiosHttpsInstance();
+    const operation = (await axios.post(`${remote}/1.0/images`, image.buffer, {
+      headers: {
+        'X-LXD-Public': '1',
+      },
+      maxContentLength: Infinity
+    })).data.operation;
```
Comment (BrunnerLivio) [inline packages/api/src/lxd/lxd.service.ts]: Can also be written as

```ts
const { data: { operation } } = await axios.post(...)
```

File: packages/api/src/lxd/lxd.service.ts (hunk 1 above)
Comment (randombenj) [reply packages/api/src/lxd/lxd.service.ts]: Ah of course this is c# eeehhr typescript :D

Comment (randombenj) [reply packages/api/src/lxd/lxd.service.ts]: Hmm but at the moment we use typescript 3.5, should we upgrade?

File: packages/api/src/cli/commands/start/default.ts [hunk 4]
```diff
@@ -88,7 +93,8 @@ export default class extends Command {
                 cert: fs.readFileSync(options.cert || `${process.env.HOME}/.config/lxc/client.crt`),
                 // @ts-ignore
                 key: fs.readFileSync(options.key || `${process.env.HOME}/.config/lxc/client.key`)
-            }
+            },
+            upload: options.upload || false
```
Comment (BrunnerLivio) [inline packages/api/src/cli/commands/start/default.ts]: If you upgraded to TypeScript 3.7, use `options.upload ?? false` instead. It is a little safer ;)

File: packages/api/src/lxd/lxd.service.ts (hunk 3 above)
Comment (randombenj) [reply packages/api/src/lxd/lxd.service.ts]: :heart_eyes:

File: packages/api/src/lxd/lxd.service.ts (hunk 1 above)
Comment (BrunnerLivio) [reply packages/api/src/lxd/lxd.service.ts]: If you can upgrade gracefully (not too much effort) then yes.

File: packages/api/src/image/image.controller.ts [hunk 5]
```diff
@@ -96,4 +107,19 @@ export class ImageController {
     : Promise<ResponseDto<CloneImageResponseDto>> {
     return await this.imageService.cloneImage(id, cloneImageDto);
   }
+
+  @Post('/')
+  @ApiResponse({ status: 200, description: 'The image was imported successfully' })
+  @UseInterceptors(FileInterceptor('image'))
+  async import(
+    @UploadedFile() image,
+    @Body('aliases') aliases: string[],
```
[the remark this answers is not in the document]
Comment (randombenj) [reply packages/api/src/image/image.controller.ts]: Doesn't work with reasonable effort ...

**Pull request merged**